src/gpu/sources.rs

```diff
@@ -11,41 +12,46 @@ static EC_SRC: &str = include_str!("multiexp/ec.cl");
 static MULTIEXP_SRC: &str = include_str!("multiexp/multiexp.cl");
 
 fn field2(field2: &str, field: &str) -> String {
-    return String::from(FIELD2_SRC)
+    String::from(FIELD2_SRC)
         .replace("FIELD2", field2)
-        .replace("FIELD", field);
+        .replace("FIELD", field)
 }
 
 fn fft(field: &str) -> String {
-    return String::from(FFT_SRC).replace("FIELD", field);
+    String::from(FFT_SRC).replace("FIELD", field)
 }
 
 fn ec(field: &str, point: &str) -> String {
-    return String::from(EC_SRC)
+    String::from(EC_SRC)
         .replace("FIELD", field)
-        .replace("POINT", point);
+        .replace("POINT", point)
 }
 
 fn multiexp(point: &str, exp: &str) -> String {
-    return String::from(MULTIEXP_SRC)
+    String::from(MULTIEXP_SRC)
         .replace("POINT", point)
-        .replace("EXPONENT", exp);
+        .replace("EXPONENT", exp)
 }
 
 // WARNING: This function works only with Short Weierstrass Jacobian curves with Fq2 extension field.
 pub fn kernel<E>() -> String
 where
     E: Engine,
 {
-    return String::from(format!(
-        "{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}",
+    let s = vec![
         ffgen::field::<E::Fr>("Fr"),
         fft("Fr"),
         ffgen::field::<E::Fq>("Fq"),
         ec("Fq", "G1"),
         multiexp("G1", "Fr"),
         field2("Fq2", "Fq"),
         ec("Fq2", "G2"),
-        multiexp("G2", "Fr")
-    ));
+        multiexp("G2", "Fr"),
+    ];
+
+    let source_code = s.join("\n\n");
+
+    debug!("Source code: \n{}", source_code);
```

Comment:
might be good to drop this to `trace`, seem that this will likely print a lot of lines

Reply:
yes, I was using this to log the actual code that was being generated. But I understand it might be too much.

Reply:
done